sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java

```diff
@@ -336,6 +342,27 @@ public Projector visitSourceIndexWriterProjection(SourceIndexWriterProjection pr
             IndexNameResolver.create(projection.tableIdent(), projection.partitionIdent(), partitionedByInputs);
         Settings tableSettings = TableSettingsResolver.get(clusterService.state().getMetaData(),
             projection.tableIdent(), !projection.partitionedBySymbols().isEmpty());
+
+
+        InputFactory.Context<CollectExpression<Row, ?>> ctxSourceInfo = inputFactory.ctxForInputColumns();
+        Input<BytesRef> sourceUriInput;
+        Input<String> sourceUriFailureInput;
+        Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> resultCollector;
+        InputColumn sourceUriSymbol = projection.sourceUriSymbol();
+        InputColumn sourceUriFailureSymbol = projection.sourceUriFailureSymbol();
+        // only take into account if both symbols are set
+        if (sourceUriSymbol != null && sourceUriFailureSymbol != null) {
+            resultCollector = UpsertResultCollectors.newSummaryCollector(clusterService.localNode());
+            //noinspection unchecked
+            sourceUriInput = (Input<BytesRef>) ctxSourceInfo.add(sourceUriSymbol);
+            //noinspection unchecked
+            sourceUriFailureInput = (Input<String>) ctxSourceInfo.add(sourceUriFailureSymbol);
+        } else {
+            resultCollector = UpsertResultCollectors.newRowCountCollector();
+            sourceUriInput = () -> null;
+            sourceUriFailureInput = () -> null;
```

Comment:
Could we extract the above if blocks into a method for clarity? Or add a comment to explain that this determines the collector behavior?

Reply:
extracting would require a transfer object as multiple vars are affected. it may not worth doing. I'll restructure this code to make it more clear, lets see if it helps...